File: Linked_List/206.Reverse_Linked_List/C++/206.Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode * reverseList(ListNode * head) 
    {
        //==============================//
        // Input type:                  //
        //   - ListNode * (head)        //
        // Return type:                 //
        //   - ListNode * (result head) //
        //==============================//

        /*Initialize*/
        ///// Next, Previous node /////
        ListNode * next_node, * prev_node = NULL;

        /*Greedy-based loop traversal*/
        while (head) //Whole
        {
            next_node = (head->next); //Next linked-list node

            (head->next) = prev_node; //Current linked-list pointer reversed
            prev_node = head; //Current stage

            head = next_node; //Next stage

        } //Whole

        return prev_node;
    }
};

File: Linked_List/206.Reverse_Linked_List/Python/206.Reverse_Linked_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        #===============================#
        # Greedy-based traversal method #
        #===============================#

        ############
        #Initialize
        ##### Previous node #####
        prev_node = None

        #############################
        #Greedy-based loop traversal
        while (head):
            next_node = (head.next) #Next node

            (head.next) = prev_node #Current head pointer reversed
            prev_node = head #Current stage

            head = next_node #Next stage 

        return prev_node
